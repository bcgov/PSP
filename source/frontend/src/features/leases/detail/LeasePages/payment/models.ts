import { defaultTypeCode } from '@/interfaces/ITypeCode';
import { ApiGen_Base_CodeType } from '@/models/api/generated/ApiGen_Base_CodeType';
import { ApiGen_Concepts_LeaseTerm } from '@/models/api/generated/ApiGen_Concepts_LeaseTerm';
import { ApiGen_Concepts_Payment } from '@/models/api/generated/ApiGen_Concepts_Payment';
import { getEmptyBaseAudit } from '@/models/defaultInitializers';
import { NumberFieldValue } from '@/typings/NumberFieldValue';
import { isValidIsoDateTime } from '@/utils';
import { stringToNumber, stringToNumberOrNull } from '@/utils/formUtils';

export class FormLeaseTerm {
  id: number | null = null;
  leaseId: number | null = null;
  statusTypeCode: ApiGen_Base_CodeType<string> | null = null;
  leasePmtFreqTypeCode: ApiGen_Base_CodeType<string> | null = null;
  startDate = '';
  effectiveDateHist: string | null = null;
  expiryDate = '';
  renewalDate = '';
  paymentAmount: NumberFieldValue = '';
  gstAmount: NumberFieldValue = '';
  paymentDueDateStr = '';
  paymentNote = '';
  isGstEligible?: boolean;
  isTermExercised?: boolean;
  payments: FormLeasePayment[] = [];
  rowVersion?: number;

  public static toApi(
    formLeaseTerm: FormLeaseTerm,
    gstConstant?: number,
  ): ApiGen_Concepts_LeaseTerm {
    return {
      ...formLeaseTerm,
      leaseId: formLeaseTerm.leaseId ?? 0,
      startDate: isValidIsoDateTime(formLeaseTerm.startDate) ? formLeaseTerm.startDate : null,
      renewalDate: null,
      expiryDate: isValidIsoDateTime(formLeaseTerm.expiryDate) ? formLeaseTerm.expiryDate : null,
      paymentAmount: stringToNumberOrNull(formLeaseTerm.paymentAmount),
      gstAmount: stringToNumberOrNull(
        formLeaseTerm.isGstEligible && gstConstant !== undefined
          ? (formLeaseTerm.paymentAmount as number) * (gstConstant / 100)
          : null,
      ),
      leasePmtFreqTypeCode: formLeaseTerm.leasePmtFreqTypeCode?.id
        ? formLeaseTerm.leasePmtFreqTypeCode
        : null,
      statusTypeCode: formLeaseTerm.statusTypeCode?.id ? formLeaseTerm.statusTypeCode : null,
      payments: formLeaseTerm.payments.map(payment => FormLeasePayment.toApi(payment)),
      isGstEligible: formLeaseTerm.isGstEligible ?? false,
      isTermExercised: formLeaseTerm.isTermExercised ?? false,
      ...getEmptyBaseAudit(formLeaseTerm.rowVersion),
    };
  }

  public static fromApi(apiLeaseTerm: ApiGen_Concepts_LeaseTerm): FormLeaseTerm {
    return {
      ...apiLeaseTerm,
      startDate: isValidIsoDateTime(apiLeaseTerm.startDate) ? apiLeaseTerm.startDate : '',
      expiryDate: isValidIsoDateTime(apiLeaseTerm.expiryDate) ? apiLeaseTerm.expiryDate : '',
      renewalDate: isValidIsoDateTime(apiLeaseTerm.renewalDate) ? apiLeaseTerm.renewalDate : '',
      paymentAmount: apiLeaseTerm.paymentAmount ?? '',
      gstAmount: apiLeaseTerm.gstAmount ?? '',
      paymentDueDateStr: apiLeaseTerm.paymentDueDateStr ?? '',
      paymentNote: apiLeaseTerm.paymentNote ?? '',
      payments:
        apiLeaseTerm.payments?.map((payment: ApiGen_Concepts_Payment) =>
          FormLeasePayment.fromApi(payment),
        ) ?? [],
      isGstEligible: apiLeaseTerm.isGstEligible ?? undefined,
      isTermExercised: apiLeaseTerm.isTermExercised ?? undefined,
      effectiveDateHist: null,
      rowVersion: apiLeaseTerm.rowVersion ?? undefined,
    };
  }
}

export const defaultFormLeaseTerm: FormLeaseTerm = {
  leaseId: 0,
  id: null,
  startDate: '',
  expiryDate: '',
  renewalDate: '',
  paymentAmount: '',
  gstAmount: '',
  paymentDueDateStr: '',
  paymentNote: '',
  isGstEligible: false,
  isTermExercised: false,
  effectiveDateHist: '',
  statusTypeCode: defaultTypeCode(),
  leasePmtFreqTypeCode: defaultTypeCode(),
  payments: [],
};

export class FormLeasePayment {
  id?: number;
  leaseTermId = 0;
  leasePaymentMethodType: ApiGen_Base_CodeType<string> | null = null;
  receivedDate = '';
  note?: string;
  leasePaymentStatusTypeCode?: ApiGen_Base_CodeType<string>;
  amountPreTax: NumberFieldValue = '';
  amountGst: NumberFieldValue = '';
  amountPst: NumberFieldValue = '';
  amountTotal: NumberFieldValue = '';
  rowVersion?: number;

  public static toApi(formLeasePayment: FormLeasePayment): ApiGen_Concepts_Payment {
    return {
      ...formLeasePayment,
      id: formLeasePayment.id ?? null,
      amountPreTax: stringToNumber(formLeasePayment.amountPreTax),
      amountGst: stringToNumber(formLeasePayment.amountGst),
      amountPst: stringToNumber(formLeasePayment.amountPst),
      amountTotal: stringToNumber(formLeasePayment.amountTotal),
      leasePaymentStatusTypeCode: formLeasePayment.leasePaymentStatusTypeCode?.id
        ? formLeasePayment.leasePaymentStatusTypeCode
        : null,
      leasePaymentMethodType: formLeasePayment.leasePaymentMethodType,
      note: formLeasePayment.note ?? null,
      ...getEmptyBaseAudit(formLeasePayment.rowVersion),
    };
  }

  public static fromApi(apiLeasePayment: ApiGen_Concepts_Payment): FormLeasePayment {
    const leasePayment = new FormLeasePayment();
    leasePayment.id = apiLeasePayment.id ?? undefined;
    leasePayment.leaseTermId = apiLeasePayment.leaseTermId;
    leasePayment.leasePaymentMethodType = apiLeasePayment.leasePaymentMethodType ?? null;
    leasePayment.receivedDate = isValidIsoDateTime(apiLeasePayment.receivedDate)
      ? apiLeasePayment.receivedDate
      : '';
    leasePayment.amountPreTax = apiLeasePayment.amountPreTax;
    leasePayment.amountPst = apiLeasePayment.amountPst ?? '';
    leasePayment.amountGst = apiLeasePayment.amountGst ?? '';
    leasePayment.amountTotal = apiLeasePayment.amountTotal ?? '';
    leasePayment.note = apiLeasePayment.note ?? undefined;
    leasePayment.leasePaymentStatusTypeCode =
      apiLeasePayment.leasePaymentStatusTypeCode ?? undefined;
    leasePayment.rowVersion = apiLeasePayment.rowVersion ?? undefined;
    return leasePayment;
  }
}

export const defaultFormLeasePayment: FormLeasePayment = {
  id: 0,
  leaseTermId: 0,
  leasePaymentMethodType: defaultTypeCode(),
  receivedDate: '',
  amountPreTax: '',
  amountPst: '',
  amountGst: '',
  amountTotal: '',
  note: '',
  leasePaymentStatusTypeCode: defaultTypeCode(),
};
