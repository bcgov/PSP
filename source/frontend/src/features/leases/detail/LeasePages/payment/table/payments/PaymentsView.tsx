import { useMemo } from 'react';
import { FaPlus } from 'react-icons/fa';
import styled from 'styled-components';

import { Section } from '@/components/common/Section/Section';
import { SectionListHeader } from '@/components/common/SectionListHeader';
import { TableProps } from '@/components/Table';
import Claims from '@/constants/claims';
import { ApiGen_CodeTypes_LeasePaymentCategoryTypes } from '@/models/api/generated/ApiGen_CodeTypes_LeasePaymentCategoryTypes';

import {
  defaultFormLeasePayment,
  FormLeasePayment,
  FormLeasePeriod,
  FormLeasePeriodWithCategory,
} from '../../models';
import * as PaymentStyles from '../../styles';
import { getLeaseVariablePeriodColumns } from '../periods/columns';
import { getActualsColumns } from './paymentsColumns';

export interface IPaymentsViewProps {
  onEdit: (values: FormLeasePayment) => void;
  onDelete: (values: FormLeasePayment) => void;
  onSave: (values: FormLeasePayment) => void;
  isExercised?: boolean;
  isGstEligible?: boolean;
  isReceivable?: boolean;
  periodId?: number;
  period: FormLeasePeriod;
}

export const PaymentsView: React.FunctionComponent<React.PropsWithChildren<IPaymentsViewProps>> = ({
  onEdit,
  onDelete,
  onSave,
  isExercised,
  isGstEligible,
  isReceivable,
  period,
}) => {
  const variablePaymentColumns = getLeaseVariablePeriodColumns();
  const variablePaymentData: FormLeasePeriodWithCategory[] = [
    { category: ApiGen_CodeTypes_LeasePaymentCategoryTypes.BASE, ...period },
    {
      category: ApiGen_CodeTypes_LeasePaymentCategoryTypes.ADDL,
      ...period,
      isGstEligible: period?.isAdditionalRentGstEligible,
      paymentAmount: period?.additionalRentPaymentAmount ?? 0,
      leasePmtFreqTypeCode: period?.additionalRentFreqTypeCode,
      gstAmount: period?.additionalRentGstAmount,
    },
    {
      category: ApiGen_CodeTypes_LeasePaymentCategoryTypes.VBL,
      ...period,
      isGstEligible: period?.isVariableRentGstEligible,
      paymentAmount: period?.variableRentPaymentAmount ?? 0,
      leasePmtFreqTypeCode: period?.variableRentFreqTypeCode,
      gstAmount: period?.variableRentGstAmount,
    },
  ];
  const columns = useMemo(
    () =>
      getActualsColumns({
        onEdit,
        onDelete,
        isReceivable,
        isGstEligible,
        onSave,
        payments: period?.payments ?? [],
      }),
    [onEdit, onDelete, isReceivable, isGstEligible, onSave, period],
  );

  return (
    <>
      {period?.isVariable === 'true' && (
        <Section className="ml-10 p-0">
          <PaymentStyles.StyledPaymentTable<React.FC<TableProps<FormLeasePeriodWithCategory>>>
            name="variablePeriodTable"
            columns={variablePaymentColumns}
            data={variablePaymentData ?? []}
            manualPagination
            hideToolbar
          />
        </Section>
      )}
      <Section
        className="ml-20 p-0"
        header={
          <StyledSectionListHeader
            title="Payments"
            addButtonText="Add a Payment"
            onAdd={
              isExercised
                ? () => onEdit({ ...defaultFormLeasePayment, leasePeriodId: period?.id ?? 0 })
                : undefined
            }
            claims={[Claims.LEASE_EDIT]}
            addButtonIcon={<FaPlus size={'2rem'} />}
          />
        }
      >
        {!!period?.payments?.length && isExercised ? (
          <>
            <PaymentStyles.StyledPaymentTable<React.FC<TableProps<FormLeasePayment>>>
              name="paymentsTable"
              columns={columns}
              data={period?.payments ?? []}
              manualPagination
              hideToolbar
              noRowsMessage="There is no corresponding data"
              footer={true}
            />
          </>
        ) : (
          <PaymentStyles.WarningTextBox>
            {!period?.payments?.length && <p>There are no recorded payments for this period.</p>}
            {!isExercised && <p>Period must be exercised to add payments.</p>}
          </PaymentStyles.WarningTextBox>
        )}
      </Section>
    </>
  );
};

export default PaymentsView;

const StyledSectionListHeader = styled(SectionListHeader)`
  font-size: 1.8rem;
`;
