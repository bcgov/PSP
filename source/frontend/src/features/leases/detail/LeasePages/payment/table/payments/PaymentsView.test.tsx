import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { Formik } from 'formik';
import { createMemoryHistory } from 'history';
import noop from 'lodash/noop';

import { Claims, LeasePeriodStatusTypes } from '@/constants';
import { LeaseFormModel } from '@/features/leases/models';
import { toTypeCodeNullable } from '@/utils/formUtils';
import { act, fillInput, renderAsync, RenderOptions, screen, userEvent } from '@/utils/test-utils';
import { getAllByRole as getAllByRoleBase } from '@/utils/test-utils';

import { defaultFormLeasePayment, defaultFormLeasePeriod, FormLeasePayment } from '../../models';
import PaymentsView, { IPaymentsViewProps } from './PaymentsView';

const history = createMemoryHistory();
const mockAxios = new MockAdapter(axios);
export const defaultTestFormLeasePayment: FormLeasePayment = {
  ...defaultFormLeasePayment,
  leasePaymentMethodType: toTypeCodeNullable('Cheque'),
  leasePaymentStatusTypeCode: toTypeCodeNullable('Paid') ?? undefined,
  amountTotal: 1200,
  amountGst: 100,
  amountPreTax: 1100,
  receivedDate: '2020-01-01',
  note: 'note',
  id: 1,
};

const getDefaultLeaseWithPeriodsPayments = () => ({
  ...new LeaseFormModel(),
  periods: [
    {
      ...defaultFormLeasePeriod,
      statusTypeCode: toTypeCodeNullable(LeasePeriodStatusTypes.EXERCISED),
      payments: [{ ...defaultTestFormLeasePayment }],
    },
  ],
});

const onEdit = vi.fn();
const onDelete = vi.fn();
const onSave = vi.fn();

describe('PaymentsView component', () => {
  const setup = async (
    renderOptions: RenderOptions &
      Partial<IPaymentsViewProps> & {
        initialValues?: any;
        onCancel?: () => void;
      } = {},
  ) => {
    // render component under test
    const component = await renderAsync(
      <PaymentsView
        onEdit={onEdit}
        onDelete={onDelete}
        onSave={onSave}
        period={renderOptions.initialValues?.periods?.[0]}
        isExercised={renderOptions?.isExercised ?? true}
        isReceivable={renderOptions?.isReceivable ?? true}
        isGstEligible={renderOptions?.isGstEligible ?? true}
      />,
      {
        ...renderOptions,
        claims: renderOptions.claims ?? [Claims.LEASE_EDIT],
        history,
      },
    );

    return {
      findFirstRow: () => {
        const rows = component.getAllByRole('row');
        return rows && rows.length > 1 ? rows[1] : null;
      },
      findFooter: () => {
        const rows = component.getAllByRole('row');
        return rows && rows.length > 1 ? rows[rows.length - 1] : null;
      },
      findCell: (row: HTMLElement, index: number) => {
        const columns = getAllByRoleBase(row, 'cell');
        return columns && columns.length > index ? columns[index] : null;
      },
      findFooterCell: (row: HTMLElement, index: number) => {
        const columns = getAllByRoleBase(row, 'columnheader');
        return columns && columns.length > index ? columns[index] : null;
      },
      component,
    };
  };

  beforeEach(() => {
    mockAxios.resetHistory();
    onEdit.mockClear();
    onSave.mockClear();
    onDelete.mockClear();
  });
  it('renders as expected', async () => {
    const { component } = await setup({});

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('renders with data as expected', async () => {
    const { component } = await setup({
      initialValues: { ...new LeaseFormModel(), periods: [defaultFormLeasePeriod] },
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe('claim behaviour', () => {
    it('Does not display edit payment if missing claim ', async () => {
      const {
        component: { queryByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        claims: [Claims.LEASE_DELETE],
      });

      const editButton = await queryByTitle('edit payment');
      expect(editButton).toBeNull();
    });

    it('Does display edit payment if has claim ', async () => {
      const {
        component: { findAllByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        claims: [Claims.LEASE_EDIT],
      });

      const editButton = await findAllByTitle('edit payment');
      expect(editButton[0]).toBeVisible();
    });

    it('Does not display delete payment if missing claim', async () => {
      const {
        component: { queryByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        claims: [],
      });

      const deleteButton = await queryByTitle('delete actual');
      expect(deleteButton).toBeNull();
    });

    it('Does display delete payment if has claim ', async () => {
      const {
        component: { findAllByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        claims: [Claims.LEASE_EDIT],
      });
      const deleteButton = await findAllByTitle('delete payment');
      expect(deleteButton[0]).toBeVisible();
    });
  });

  describe('view only logic', () => {
    it('Displays there are no payments message if there are no payments', async () => {
      const {
        component: { findByText },
      } = await setup({
        initialValues: {
          ...getDefaultLeaseWithPeriodsPayments(),
          periods: [{ ...defaultFormLeasePeriod }],
        },
      });
      const text = await findByText('There are no recorded payments for this period.');
      expect(text).toBeVisible();
    });

    it('Displays not exercised message if period is not exercised', async () => {
      const {
        component: { findByText },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isExercised: false,
      });
      const text = await findByText('Period must be exercised to add payments.');
      expect(text).toBeVisible();
    });

    it('Displays proper received columns if received', async () => {
      const {
        component: { findByText },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isReceivable: true,
      });
      expect(await findByText('Payments')).toBeVisible();
      expect(await findByText('Date')).toBeVisible();
      expect(await findByText('Received payment ($)', { exact: false })).toBeVisible();
      expect(await findByText('Total ($)')).toBeVisible();
    });

    it('Displays proper sent columns if payable', async () => {
      const {
        component: { findByText },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isReceivable: false,
      });
      expect(await findByText('Payments')).toBeVisible();
      expect(await findByText('Date')).toBeVisible();
      expect(await findByText('Sent payment ($)')).toBeVisible();
      expect(await findByText('Total ($)')).toBeVisible();
    });

    it('Does not display GST values or calculations if period is not gst eligible', async () => {
      const { findFooter, findCell, findFooterCell, findFirstRow } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isGstEligible: false,
      });
      const row = findFirstRow() as HTMLElement;
      const footer = findFooter() as HTMLElement;
      expect(row).not.toBeNull();
      expect(footer).not.toBeNull();
      expect(findCell(row, 4)?.textContent).toBe('-');
      expect(findFooterCell(footer, 4)?.textContent).toBe('-');
    });

    it('Sums expected payment columns', async () => {
      const { findFooter, findCell, findFooterCell, findFirstRow } = await setup({
        initialValues: {
          ...new LeaseFormModel(),
          periods: [
            {
              ...defaultFormLeasePeriod,
              payments: [defaultTestFormLeasePayment, defaultTestFormLeasePayment],
            },
          ],
        },
      });
      const row = findFirstRow() as HTMLElement;
      const footer = findFooter() as HTMLElement;
      expect(row).not.toBeNull();
      expect(footer).not.toBeNull();
      expect(findFooterCell(footer, 2)?.textContent).toBe('(2) payments');
      expect(findCell(row, 3)?.textContent).toBe('$1,100.00');
      expect(findFooterCell(footer, 3)?.textContent).toBe('$2,200.00');
      expect(findCell(row, 4)?.textContent).toBe('$100.00');
      expect(findFooterCell(footer, 4)?.textContent).toBe('$200.00');
      expect(findCell(row, 5)?.textContent).toBe('$1,200.00');
      expect(findFooterCell(footer, 5)?.textContent).toBe('$2,400.00');
    });
  });
  describe('interactive functionality', () => {
    it('Allows user to enter and save a note', async () => {
      const {
        component: { findByTitle, getByText },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isReceivable: false,
      });
      const notesButton = await findByTitle('Payment comments');
      await act(async () => {
        userEvent.click(notesButton);
      });
      await act(async () => {
        await fillInput(document.body, 'note', 'a test note', 'textarea');
      });
      const saveButton = getByText('Yes');
      await act(async () => {
        userEvent.click(saveButton);
      });
      expect(onSave).toHaveBeenCalledWith({ ...defaultTestFormLeasePayment, note: 'a test note' });
    });

    it('Allows user to trigger an edit action', async () => {
      const {
        component: { findAllByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isReceivable: false,
      });
      const editButton = await findAllByTitle('edit payment');
      await act(async () => userEvent.click(editButton[0]));
      expect(onEdit).toHaveBeenCalledWith(defaultTestFormLeasePayment);
    });

    it('Allows user to trigger a delete action', async () => {
      const {
        component: { findAllByTitle },
      } = await setup({
        initialValues: getDefaultLeaseWithPeriodsPayments(),
        isReceivable: false,
        claims: [Claims.LEASE_EDIT],
      });
      const deleteButton = await findAllByTitle('delete payment');
      await act(async () => userEvent.click(deleteButton[0]));
      expect(onDelete).toHaveBeenCalledWith(defaultTestFormLeasePayment);
    });
  });
});
