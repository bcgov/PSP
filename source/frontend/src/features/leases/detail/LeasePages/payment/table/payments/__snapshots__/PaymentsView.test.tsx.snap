// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PaymentsView component > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c4.btn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.4rem 1.2rem;
  min-height: 3rem;
  border: 0.2rem solid transparent;
  border-radius: 0.4rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.8rem;
  font-family: 'BCSans','Noto Sans',Verdana,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  cursor: pointer;
  height: 3.8rem;
}

.c4.btn .Button__value {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c4.btn:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.8;
}

.c4.btn:focus {
  outline-width: 0.4rem;
  outline-style: solid;
  outline-offset: 1px;
  box-shadow: none;
}

.c4.btn.btn-primary {
  border: none;
}

.c4.btn.btn-secondary {
  background: none;
}

.c4.btn.btn-info {
  border: none;
  background: none;
  padding-left: 0.6rem;
  padding-right: 0.6rem;
}

.c4.btn.btn-info:hover,
.c4.btn.btn-info:active,
.c4.btn.btn-info:focus {
  background: none;
}

.c4.btn.btn-light {
  border: none;
}

.c4.btn.btn-dark {
  border: none;
}

.c4.btn.btn-link {
  font-size: 1.6rem;
  font-weight: 400;
  background: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 2.5rem;
  line-height: 3rem;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-letter-spacing: unset;
  -moz-letter-spacing: unset;
  -ms-letter-spacing: unset;
  letter-spacing: unset;
  text-align: left;
  padding: 0;
}

.c4.btn.btn-link:hover,
.c4.btn.btn-link:active,
.c4.btn.btn-link:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border: none;
  background: none;
  box-shadow: none;
  outline: none;
}

.c4.btn.btn-link:disabled,
.c4.btn.btn-link.disabled {
  background: none;
  pointer-events: none;
}

.c4.btn:disabled,
.c4.btn:disabled:hover {
  box-shadow: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.65;
}

.c4.Button .Button__icon {
  margin-right: 1.6rem;
}

.c4.Button--icon-only:focus {
  outline: none;
}

.c4.Button--icon-only .Button__icon {
  margin-right: 0;
}

.c1 {
  font-weight: bold;
  border-bottom: 0.2rem solid;
  margin-bottom: 2rem;
}

.c0 {
  margin: 1.6rem;
  padding: 1.6rem;
  background-color: white;
  text-align: left;
  border-radius: 0.5rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c5.c5 {
  display: inline-block;
  margin-left: 1.5rem;
  margin-bottom: 0.5rem;
}

.c2 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  min-height: 4.5rem;
}

.c2 .btn {
  margin: 0;
}

.c7 {
  gap: 3rem;
  background-color: white;
  min-height: 6rem;
}

.c3 {
  font-size: 1.8rem;
}

<div
    class="c0 form-section ml-20 p-0"
  >
    <h2
      class="c1"
    >
      <div
        class="no-gutters row"
      >
        <div
          class="col"
        >
          <div
            class="c2 no-gutters c3 row"
          >
            <div
              class="px-2 my-1 col-auto"
            >
              Payments
            </div>
            <div
              class="my-1 col-auto"
            >
              <button
                class="c4 Button c5 btn btn-primary"
                type="button"
              >
                <div
                  class="Button__value"
                >
                  <svg
                    fill="currentColor"
                    height="2rem"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="2rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                  Â Add a Payment
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </h2>
    <div
      class="collapse show"
    >
      <div
        class="c6 c7"
      >
        <p>
          There are no recorded payments for this period.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PaymentsView component > renders with data as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c4.btn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.4rem 1.2rem;
  min-height: 3rem;
  border: 0.2rem solid transparent;
  border-radius: 0.4rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.8rem;
  font-family: 'BCSans','Noto Sans',Verdana,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  cursor: pointer;
  height: 3.8rem;
}

.c4.btn .Button__value {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c4.btn:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.8;
}

.c4.btn:focus {
  outline-width: 0.4rem;
  outline-style: solid;
  outline-offset: 1px;
  box-shadow: none;
}

.c4.btn.btn-primary {
  border: none;
}

.c4.btn.btn-secondary {
  background: none;
}

.c4.btn.btn-info {
  border: none;
  background: none;
  padding-left: 0.6rem;
  padding-right: 0.6rem;
}

.c4.btn.btn-info:hover,
.c4.btn.btn-info:active,
.c4.btn.btn-info:focus {
  background: none;
}

.c4.btn.btn-light {
  border: none;
}

.c4.btn.btn-dark {
  border: none;
}

.c4.btn.btn-link {
  font-size: 1.6rem;
  font-weight: 400;
  background: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 2.5rem;
  line-height: 3rem;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-letter-spacing: unset;
  -moz-letter-spacing: unset;
  -ms-letter-spacing: unset;
  letter-spacing: unset;
  text-align: left;
  padding: 0;
}

.c4.btn.btn-link:hover,
.c4.btn.btn-link:active,
.c4.btn.btn-link:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border: none;
  background: none;
  box-shadow: none;
  outline: none;
}

.c4.btn.btn-link:disabled,
.c4.btn.btn-link.disabled {
  background: none;
  pointer-events: none;
}

.c4.btn:disabled,
.c4.btn:disabled:hover {
  box-shadow: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.65;
}

.c4.Button .Button__icon {
  margin-right: 1.6rem;
}

.c4.Button--icon-only:focus {
  outline: none;
}

.c4.Button--icon-only .Button__icon {
  margin-right: 0;
}

.c1 {
  font-weight: bold;
  border-bottom: 0.2rem solid;
  margin-bottom: 2rem;
}

.c0 {
  margin: 1.6rem;
  padding: 1.6rem;
  background-color: white;
  text-align: left;
  border-radius: 0.5rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c5.c5 {
  display: inline-block;
  margin-left: 1.5rem;
  margin-bottom: 0.5rem;
}

.c2 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  min-height: 4.5rem;
}

.c2 .btn {
  margin: 0;
}

.c7 {
  gap: 3rem;
  background-color: white;
  min-height: 6rem;
}

.c3 {
  font-size: 1.8rem;
}

<div
    class="c0 form-section ml-20 p-0"
  >
    <h2
      class="c1"
    >
      <div
        class="no-gutters row"
      >
        <div
          class="col"
        >
          <div
            class="c2 no-gutters c3 row"
          >
            <div
              class="px-2 my-1 col-auto"
            >
              Payments
            </div>
            <div
              class="my-1 col-auto"
            >
              <button
                class="c4 Button c5 btn btn-primary"
                type="button"
              >
                <div
                  class="Button__value"
                >
                  <svg
                    fill="currentColor"
                    height="2rem"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="2rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                  Â Add a Payment
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </h2>
    <div
      class="collapse show"
    >
      <div
        class="c6 c7"
      >
        <p>
          There are no recorded payments for this period.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
